src/madness/chem/lowrankfunction.h

```diff
@@ -0,0 +1,587 @@
+                cartesian_grid<LDIM> cg(nperdim,-radius,radius);
+                for (cg.index=0; cg(); ++cg) grid.push_back(cg.get_coordinates());
+                if (world.rank()==0 and do_print) print("volume element in cartesian grid",cg.volume_per_gridpoint());
+            } else {
+                MADNESS_EXCEPTION("unknown grid type in project",1);
+            }
+
+            if (world.rank()==0 and do_print) {
+                print("grid is",gridtype,"with radius",radius,"and",grid.size(),"gridpoints");
+                print("rhs functions are",rhsfunctiontype);
+            }
+
+            auto Y=Yformer(grid,rhsfunctiontype);
+            t1.tag("Yforming");
+
+            std::ostringstream oss;
+            oss << std::scientific << std::setprecision(1) << rank_revealing_tol;
+            std::string scientificString = oss.str();
+            g=orthonormalize_rrcd(Y,rank_revealing_tol);
+            t1.tag("Y orthonormalizing with rank_revealing_tol "+scientificString);
+
+            if (world.rank()==0 and do_print) {
+                print("Y.size()",Y.size());
+                print("g.size()",g.size());
+            }
+
+            h=truncate(inner(lrfunctor,g,p1,p1));
+            t1.tag("Y backprojection with truncation");
+
+        }
+
+        /// apply a rhs (delta or exponential) on grid points to the hi-dim function and form Y = A_ij w_j (in Halko's language)
+        std::vector<Function<T,LDIM>> Yformer(const std::vector<Vector<double,LDIM>>& grid, const std::string rhsfunctiontype,
+                                              const double exponent=30.0) {
+
+            std::vector<Function<double,LDIM>> Y;
+            if (rhsfunctiontype=="delta") {
+                Y=inner(lrfunctor,grid,p2,p1);
+
+            } else if (rhsfunctiontype=="exponential") {
+                std::vector<Function<double,LDIM>> omega;
+                for (const auto& point : grid) {
+                    omega.push_back(FunctionFactory<double,LDIM>(world)
+                            .functor([&point,&exponent](const Vector<double,LDIM>& r)
+                                     {
+                                         auto r_rel=r-point;
+                                         return exp(-exponent*madness::inner(r_rel,r_rel));
+                                     }));
+                }
+                Y=inner(lrfunctor,omega,p2,p1);
+            } else {
+                MADNESS_EXCEPTION("confused rhsfunctiontype",1);
+            }
+            return Y;
+        }
+
+        long rank() const {return g.size();}
+
+        Function<T,NDIM> reconstruct() const {
+            auto fapprox=hartree_product(g[0],h[0]);
+            for (int i=1; i<g.size(); ++i) fapprox+=hartree_product(g[i],h[i]);
+            return fapprox;
+        }
+
+        /// optimize using Cholesky decomposition
+
+        /// if stable_power_iteration is true, orthonormalize in between applications of the kernel (Alg. 4.4 in Halko)
```

Comment:
This one: https://arxiv.org/abs/0909.4061 ?
a bit hard to follow what is happening here.